File: android/lib/src/main/java/com/ernpetstore/ern/api/StoreRequests.java
/*
* Copyright 2017 WalmartLabs
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
* http://www.apache.org/licenses/LICENSE-2.0
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

package com.ernpetstore.ern.api;

import android.support.annotation.NonNull;

import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeHolder;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeRequestHandler;
import com.walmartlabs.electrode.reactnative.bridge.ElectrodeBridgeResponseListener;
import com.walmartlabs.electrode.reactnative.bridge.None;
import com.walmartlabs.electrode.reactnative.bridge.RequestHandlerProcessor;
import com.walmartlabs.electrode.reactnative.bridge.RequestProcessor;
import java.util.*;
import com.ernpetstore.ern.model.Order;
import java.util.Map;


final class StoreRequests implements StoreApi.Requests {
    StoreRequests() {}


    @Override
    public void registerDeleteOrderRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<String, None> handler) {
        new RequestHandlerProcessor<>(REQUEST_DELETE_ORDER, String.class, None.class, handler).execute();
    }

    @Override
    public void registerGetInventoryRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<None, Map<String, Integer>> handler) {
        new RequestHandlerProcessor<>(REQUEST_GET_INVENTORY, None.class, Map<String, Integer>.class, handler).execute();
    }

    @Override
    public void registerGetOrderByIdRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<String, Order> handler) {
        new RequestHandlerProcessor<>(REQUEST_GET_ORDER_BY_ID, String.class, Order.class, handler).execute();
    }

    @Override
    public void registerPlaceOrderRequestHandler(@NonNull final ElectrodeBridgeRequestHandler<Order, Order> handler) {
        new RequestHandlerProcessor<>(REQUEST_PLACE_ORDER, Order.class, Order.class, handler).execute();
    }

    //------------------------------------------------------------------------------------------------------------------------------------

    @Override
    public void deleteOrder(String orderId,@NonNull final ElectrodeBridgeResponseListener<None> responseListener) {
        new RequestProcessor<>(REQUEST_DELETE_ORDER,  orderId, None.class, responseListener).execute();
    }
    @Override
    public void getInventory(@NonNull final ElectrodeBridgeResponseListener<Map<String, Integer>> responseListener) {
        new RequestProcessor<>(REQUEST_GET_INVENTORY, null, Map<String, Integer>.class, responseListener).execute();
    }
    @Override
    public void getOrderById(String orderId,@NonNull final ElectrodeBridgeResponseListener<Order> responseListener) {
        new RequestProcessor<>(REQUEST_GET_ORDER_BY_ID,  orderId, Order.class, responseListener).execute();
    }
    @Override
    public void placeOrder(Order body,@NonNull final ElectrodeBridgeResponseListener<Order> responseListener) {
        new RequestProcessor<>(REQUEST_PLACE_ORDER,  body, Order.class, responseListener).execute();
    }
}

File: IOS/StoreRequests.swift

public class StoreRequests: StoreAPIRequests {

    public override func registerDeleteOrderRequestHandler(handler:  @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: StoreAPI.kRequestDeleteOrder,
    reqClass: String.self, 
    respClass: None.self,
    requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    public override func registerGetInventoryRequestHandler(handler:  @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: StoreAPI.kRequestGetInventory,
    reqClass: None.self, 
    respClass: [String:Int].self,
    requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    public override func registerGetOrderByIdRequestHandler(handler:  @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: StoreAPI.kRequestGetOrderById,
    reqClass: String.self, 
    respClass: Order.self,
    requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    public override func registerPlaceOrderRequestHandler(handler:  @escaping ElectrodeBridgeRequestCompletionHandler) {
        let requestHandlerProcessor = ElectrodeRequestHandlerProcessor(requestName: StoreAPI.kRequestPlaceOrder,
    reqClass: Order.self, 
    respClass: Order.self,
    requestCompletionHandler: handler)
        requestHandlerProcessor.execute()
    }

    //------------------------------------------------------------------------------------------------------------------------------------


    public override func deleteOrder(orderId: String, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<String, None, Any>(
            requestName: StoreAPI.kRequestDeleteOrder,
            requestPayload: orderId,
            respClass: None.self,
            responseItemType: nil,
            responseCompletionHandler: responseCompletionHandler)

        requestProcessor.execute()
    }

    public override func getInventory( responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<None, [String:Int], Any>(
            requestName: StoreAPI.kRequestGetInventory,
            requestPayload: nil, 
            respClass: [String:Int].self,
            responseItemType: nil,
            responseCompletionHandler: responseCompletionHandler)

        requestProcessor.execute()
    }

    public override func getOrderById(orderId: String, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<String, Order, Any>(
            requestName: StoreAPI.kRequestGetOrderById,
            requestPayload: orderId,
            respClass: Order.self,
            responseItemType: nil,
            responseCompletionHandler: responseCompletionHandler)

        requestProcessor.execute()
    }

    public override func placeOrder(body: Order, responseCompletionHandler: @escaping ElectrodeBridgeResponseCompletionHandler) {
        let requestProcessor = ElectrodeRequestProcessor<Order, Order, Any>(
            requestName: StoreAPI.kRequestPlaceOrder,
            requestPayload: body,
            respClass: Order.self,
            responseItemType: nil,
            responseCompletionHandler: responseCompletionHandler)

        requestProcessor.execute()
    }
}